core/src/block.rs

```diff
@@ -363,30 +361,50 @@ pub struct ValidBlock {
     pub rejected_transactions: Vec<VersionedRejectedTransaction>,
     /// Array of all transactions in this block.
     pub transactions: Vec<VersionedValidTransaction>,
-    /// Signatures of peers which approved this block.
-    pub signatures: BTreeSet<SignatureOf<Self>>,
     /// Event recommendations.
     pub event_recommendations: Vec<Event>,
 }
 
 impl ValidBlock {
+    /// Calculate hash of the current block.
+    pub fn hash(&self) -> HashOf<Self> {
+        HashOf::new(&self.header).transmute()
+    }
+
+    /// Sign this block and get `ValidSignedBlock`.
+    ///
     /// # Errors
-    /// Asserts specific transaction limits hold true
-    pub fn check_transaction_limits(&self, tx_limits: &TransactionLimits) -> Result<()> {
-        self.transactions
-            .iter()
-            .map(|tx| tx.check_limits(tx_limits))
-            .collect::<Result<Vec<_>, _>>()
-            .map(drop)?;
-        // TODO: Check Event recommendations.
-        self.rejected_transactions
-            .iter()
-            .map(|tx| tx.check_limits(tx_limits))
-            .collect::<Result<Vec<_>, _>>()
-            .map(drop)?;
-        Ok(())
+    /// Fails if generating signature fails
+    pub fn sign(self, key_pair: KeyPair) -> Result<ValidSignedBlock> {
+        let signature = SignatureOf::from_hash(key_pair, &self.hash().transmute())
+            .wrap_err("Failed to sign block")?;
+        let signatures = SignaturesOf::from(signature);
+        Ok(ValidSignedBlock {
+            header: self.header,
+            rejected_transactions: self.rejected_transactions,
+            transactions: self.transactions,
+            event_recommendations: self.event_recommendations,
+            signatures,
+        })
     }
+}
 
+/// After receiving first signature `ValidBlock` can transform into `ValidSignedBlock`.
+#[derive(Debug, Clone)]
+pub struct ValidSignedBlock {
+    /// Header
+    pub header: BlockHeader,
+    /// Array of rejected transactions.
+    pub rejected_transactions: Vec<VersionedRejectedTransaction>,
+    /// Array of transactions.
+    pub transactions: Vec<VersionedValidTransaction>,
+    /// Signatures of peers which approved this block.
+    pub signatures: SignaturesOf<Self>,
+    /// Event recommendations.
+    pub event_recommendations: Vec<Event>,
+}
+
+impl ValidSignedBlock {
     /// Commit block to the store.
     pub fn commit(self) -> CommittedBlock {
```

Comment:
If we don't use this function outside this crate, I suggest making it `pub(crate)`. Alternatively, if we want to keep it in the public API, consider making it `#[inline]`.

Reply:
Made `inline`.